cache/save_skip.go

```diff
@@ -1,48 +1,100 @@
+		return "new_archive_checksum_match"
+	case reasonNewArchiveChecksumMismatch:
+		return "new_archive_checksum_mismatch"
+	default:
+		return "unknown"
+	}
+}
+
+func (r skipReason) description() string {
+	switch r {
+	case reasonKeyNotDynamic:
+		return "key is not dynamic; the expectation is that the same key is used for saving different cache contents over and over"
+	case reasonNoRestore:
+		return "no cache was restored in the workflow, creating a new cache entry"
+	case reasonRestoreSameUniqueKey:
+		return "a cache with the same key was restored in the workflow, new cache would have the same content"
+	case reasonRestoreSameKeyNotUnique:
+		return "a cache with the same key was restored in the workflow, but contents might have changed since then"
+	case reasonRestoreOtherKey:
+		return "there was no cache restore in the workflow with this key"
+	case reasonNewArchiveChecksumMatch:
+		return "new cache archive is the same as the restored one"
+	case reasonNewArchiveChecksumMismatch:
+		return "new cache archive doesn't match the restored one"
+	default:
+		return "unrecognized skipReason"
+	}
+}
+
+func (s *saver) canSkipSave(keyTemplate, evaluatedKey string, isKeyUnique bool) (bool, skipReason) {
 	if keyTemplate == evaluatedKey {
-		return false, "key is not dynamic; the expectation is that the same key is used for saving different cache contents over and over"
+		return false, reasonKeyNotDynamic
 	}
 
 	cacheHits := s.getCacheHits()
 	if len(cacheHits) == 0 {
-		return false, "no cache was restored in the workflow, creating a new cache entry"
+		return false, reasonNoRestore
 	}
 
 	if _, ok := cacheHits[evaluatedKey]; ok {
-		if onlyCheckCacheKey {
-			return true, "a cache with the same key was restored in the workflow, new cache would have the same content"
+		if isKeyUnique {
+			return true, reasonRestoreSameUniqueKey
 		} else {
-			return false, "a cache with the same key was restored in the workflow, but contents might have changed since then"
+			return false, reasonRestoreSameKeyNotUnique
 		}
 	}
 
-	otherKeys := []string{}
-	for k := range cacheHits {
-		otherKeys = append(otherKeys, k)
-	}
-	fullReason := fmt.Sprintf("there was no cache restore in the workflow with this key\nOther restored cache keys found:\n%s", strings.Join(otherKeys, "\n"))
-	return false, fullReason
+	return false, reasonRestoreOtherKey
 }
 
-func (s *saver) canSkipUpload(newCacheKey, newCacheChecksum string) (canSkip bool, reason string) {
+func (s *saver) canSkipUpload(newCacheKey, newCacheChecksum string) (bool, skipReason) {
 	cacheHits := s.getCacheHits()
 
 	if len(cacheHits) == 0 {
-		return false, "no cache was restored in the workflow"
+		return false, reasonNoRestore
 	}
 
-	if cacheHits[newCacheKey] == newCacheChecksum {
-		return true, "new cache archive is the same as the restored one"
-	} else {
-		return false, "new cache archive doesn't match the restored one"
+	checksumForNewKey, ok := cacheHits[newCacheKey]
```

Comment:
There was a subtle mistake in the old implementation, in reality there are 3 distinct cases, not 2.